src/server/transaction.cc

```diff
@@ -1060,26 +1077,25 @@ bool Transaction::ScheduleInShard(EngineShard* shard, bool execute_optimistic) {
     // Check if we can run immediately
     if (shard_unlocked && execute_optimistic &&
         CheckLocks(GetDbSlice(shard->shard_id()), mode, lock_args)) {
+      // We need to acquire the fp locks because the executing callback
+      // within RunCallback might preempt.
+      acquire_fp_locks(shard_unlocked);
       sd.local_mask |= OPTIMISTIC_EXECUTION;
       shard->stats().tx_optimistic_total++;
 
       RunCallback(shard);
 
       // Check state again, it could've been updated if the callback returned AVOID_CONCLUDING flag.
       // Only possible for single shard.
-      if (coordinator_state_ & COORD_CONCLUDING)
+      if (coordinator_state_ & COORD_CONCLUDING) {
```

Comment:
@dranikpg At first I was confused by this because for `execute_optimistic` to be `true` we need `coordinator_state_ & COORD_CONCLUDING`. However, I found out by looking at the stream family, that there are transactions that start with this state and then they might `AVOID_CONCLUDING` depending on the callback result. I almost removed the `if` here :smile:

Reply:
> However, I found out by looking at the stream family,

You could've just looked at the comment above 😂 I don't write them just for fun